third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -0,0 +1,286 @@
+    constOffsets.push_back(
+        builder.create<arith::ConstantIntOp>(loc, offset, 32));
+  }
+}
+
+LogicalResult Pingponger::genLocalSlice(OpBuilder &builder, Value v,
+                                        Attribute dotEncoding, unsigned opIdx,
+                                        unsigned numSlices,
+                                        int64_t sliceWidth) {
+  SmallVector<Operation *> slices;
+  SmallVector<Operation *> subviews;
+  auto memDesc = v.getDefiningOp()->getOperand(0);
+  auto type = cast<ttg::MemDescType>(memDesc.getType());
+  auto encoding = cast<RankedTensorType>(v.getType()).getEncoding();
+  auto srcEncoding = mlir::cast<ttg::DotOperandEncodingAttr>(encoding);
+  SmallVector<int64_t> shape{type.getShape().begin(), type.getShape().end()};
+  Type elementType = type.getElementType();
+  int64_t kIdx = opIdx == 0 ? 1 : 0;
+  shape[kIdx] = sliceWidth;
+  if (sliceWidth < 16)
+    return failure();
+
+  for (int i = 0; i < numSlices; i++) {
+    SmallVector<Value> offsetsVal;
+    SmallVector<int64_t> offsets = {0, 0};
+    offsets[kIdx] = i;
+    for (int64_t off : offsets) {
+      offsetsVal.push_back(constOffsets[off]);
+    }
+
+    Value newSmem = builder.create<triton::gpu::MemDescSubviewOp>(
+        v.getLoc(),
+        ttg::MemDescType::get(shape, elementType, type.getEncoding(),
+                              type.getMemorySpace()),
+        memDesc, offsetsVal);
+    auto dotOperandEnc = triton::gpu::DotOperandEncodingAttr::get(
+        builder.getContext(), opIdx, dotEncoding, srcEncoding.getKWidth());
+    Value prefetchSlice = builder.create<triton::gpu::LocalLoadOp>(
+        v.getLoc(), RankedTensorType::get(shape, elementType, dotOperandEnc),
+        newSmem);
+    subviews.push_back(newSmem.getDefiningOp());
+    slices.push_back(prefetchSlice.getDefiningOp());
+  }
+
+  subViewOps.push_back(subviews);
+  loadSliceOps.push_back(slices);
+  return success();
+}
+
+// Transform two warps coming from the same blocks
+LogicalResult Pingponger::transformTwoWarp(OpBuilder &builder, Location loc) {
+  // First, slice local_loads and dot into 4 parts
+  unsigned numSlices = 4;
+  auto op = cast<triton::DotOp>(dotOps[0]);
+  builder.setInsertionPointToStart(forOp.getBody());
+  auto typeB = op.getB().getType();
+  auto shapeB = typeB.getShape();
+  int64_t sliceWidth = shapeB[0] / numSlices;
+  genOffsetConstants(loc, builder, numSlices, sliceWidth);
+
+  builder.setInsertionPointAfter(gLoadOps[0]);
+  auto dotEncoding = op.getType().getEncoding();
+  if (genLocalSlice(builder, op.getA(), dotEncoding, 0, numSlices, sliceWidth)
+          .failed() ||
+      genLocalSlice(builder, op.getB(), dotEncoding, 1, numSlices, sliceWidth)
+          .failed())
+    return failure();
+
+  // Clone dots four times to consume the slices
+  Operation *prevDot = op;
+  for (int i = 0; i < numSlices; i++) {
+    auto newOp = builder.clone(*op);
+    newOp->setOperand(0, loadSliceOps[0][i]->getResult(0));
```

Comment:
Use `IRMapping` utility so you can do the remapping when `clone`?

Reply:
Oh, missed this one will do it